src/Lean/Data/Lsp/Basic.lean

```diff
@@ -0,0 +1,313 @@
+/-
+Copyright (c) 2020 Marc Huisinga. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Marc Huisinga, Wojciech Nawrocki
+-/
+import Lean.Data.Json
+
+/-! Defines most of the 'Basic Structures' in the LSP specification
+(https://microsoft.github.io/language-server-protocol/specifications/specification-current/),
+as well as some utilities.
+
+Since LSP is Json-based, Ints/Nats are represented by Floats on the wire. -/
+
+namespace Lean
+namespace Lsp
+
+open Json
+
+abbrev DocumentUri := String
+
+/-- An array of the lines in a text document.
+Elements mustn't contain newline characters. -/
+def DocumentText := Array String
+
+/-- Can represent both a UTF-16 position - this is what LSP clients send -
+and a codepoint position - what we use internally.
+`character` is accepted liberally: actual character := min(line length, character) -/
+structure Position := (line : Nat) (character : Nat)
+
+instance Position.hasFromJson : HasFromJson Position :=
+⟨fun j => do
+  line ← j.getObjValAs? Nat "line";
+  character ← j.getObjValAs? Nat "character";
+  pure ⟨line, character⟩⟩
+
+instance Position.hasToJson : HasToJson Position :=
+⟨fun o => mkObj [
+  ⟨"line", o.line⟩,
+  ⟨"character", o.character⟩]⟩
+
+instance Position.hasToString : HasToString Position :=
+⟨fun p => "(" ++ toString p.line ++ ", " ++ toString p.character ++ ")"⟩
+
+namespace DocumentText
+
+/-- Computes a linear position in `("\n".intercalate text.toList)`
+from an LSP-style 0-indexed (ln, col) position. -/
+def lnColToLinearPos (text : DocumentText) (pos : Position) : String.Pos :=
```

Comment:
Ah, then my model of it was wrong. This [comment](https://github.com/leanprover/lean4/pull/161/files#diff-77b675b836442479de606add77a90317R33) also ties in. Is there a type for codepoint offsets? Even though `String` has an `extern` implementation, I would expect the position type to refer to codepoints, because the pure-`Lean` `String` is just a list of Unicode `Char` scalars. Would it be possible to make `Pos` just a position in this list (so a codepoint offset) and perhaps add another `String.Utf8Pos`?

Reply:
I believe the original intention was for `String.Pos` to be a more or less opaque type that hides the actual encoding but guarantees constant-time access, much like `String` itself. Users probably should never do arithmetic on `String.Pos` themselves. If you want to count actual characters etc., you should simply use `Nat`. @leodemoura?